### DISCUSSION TITLE
Modified Virus_On_Network Blank Visualization in Server

### DISCUSSION THREAD

**Discussion by jenny61599** [mesa/mesa#1616] [Q&A] [answered]
I modified the Virus_On_Network example so it could read in .edge files. However, I am running into the issue of the grid and chart not showing up. Furthermore, when I click start, the steps are not going. The actual model is working though after testing it without using the server. I will show below the issue, things I have done/troubleshoot in server.py. My version of mesa is 1.1 and python 3.10. I can run the original Virus_On_Network perfectly.

**The issue:**
![Screenshot 2023-03-15 114539](https://user-images.githubusercontent.com/81499095/225412178-bd0137b3-47f8-46b6-8bff-f5ea670bf170.png)

**Things I have done on server.py:** 
1) Modify the example so network_files is part of the model parameter that reads in a path to a .edge file. Got rid of the first two sliders.
```
model_params = {
    "network_file": r"C:\Users\name\OneDrive\One Drive Desktop\0.edges",

    "initial_outbreak_size": mesa.visualization.Slider(
        "Initial Outbreak Size",
        1,
        1,
        10,
        1,
        description="Initial Outbreak Size",
    ),
    "virus_spread_chance": mesa.visualization.Slider(
        "Virus Spread Chance",
        0.4,
        0.0,
        1.0,
        0.1,
        description="Probability that susceptible neighbor will be infected",
    ),
    "virus_check_frequency": mesa.visualization.Slider(
        "Virus Check Frequency",
        0.4,
        0.0,
        1.0,
        0.1,
        description="Frequency the nodes check whether they are infected by " "a virus",
    ),
    "recovery_chance": mesa.visualization.Slider(
        "Recovery Chance",
        0.3,
        0.0,
        1.0,
        0.1,
        description="Probability that the virus will be removed",
    ),
    "gain_resistance_chance": mesa.visualization.Slider(
        "Gain Resistance Chance",
        0.5,
        0.0,
        1.0,
        0.1,
        description="Probability that a recovered agent will become "
        "resistant to this virus in the future",
    ),
}
```
2.) Test if portrayal works by print(portrayal). I get an output that looks correct.
For edges
`{'source': 77, 'target': 294, 'color': '#e8e8e8', 'width': 2}, {'source': 219, 'target': 154, 'color': '#e8e8e8', 'width': 2}, {'source': 182, 'target': 253, 'color': '#e8e8e8', 'width': 2}]}`
For nodes
`{'size': 6, 'color': '#008000', 'tooltip': 'id: 267<br>state: SUSCEPTIBLE'}, {'size': 6, 'color': '#008000', 'tooltip': 'id: 316<br>state: SUSCEPTIBLE'}, {'size': 6, 'color': '#008000', 'tooltip': 'id: 205<br>state: SUSCEPTIBLE'}`

3) Test if network works by print(network), it returns an network object so it should be working.
```
network = mesa.visualization.NetworkModule(network_portrayal, 500, 500)
print(network)
```
returns
`<mesa.visualization.modules.NetworkVisualization.NetworkModule object at 0x000002441E1F28C0>`

4. Lastly, here is my code at the very end, which is not modified.
```
server = mesa.visualization.ModularServer(
    VirusOnNetwork,
    [network, get_resistant_susceptible_ratio, chart],
    "Virus Model",
    model_params,
)
server.port = 8521
```

Anyone got an idea of what's going on?

**Comment by rht** [comment]
I used the exact same `model_params` you posted, and it rendered fine. Any error messages you can find from the browser console messages? It is likely something to do with the `network_file` handling.

**Reply by jenny61599** [reply]
Thanks for responding! What I found is that if I remove `network ` from the previous code and run this instead:
```
server = mesa.visualization.ModularServer(
    VirusOnNetwork,
    [get_resistant_susceptible_ratio, chart],
    "Virus Model",
    model_params,
)
```
The model actually runs fine with just the chart, but if I replace `[get_resistant_susceptible_ratio, chart]` with `[network]` or `[network, get_resistant_susceptible_ratio, chart]`. The result remains the same as before. 

![Screenshot 2023-03-16 011712](https://user-images.githubusercontent.com/81499095/225556160-370b897a-4468-49a4-8ad3-8f89e332fe03.png)

I don't have any browser console messages, but I can include more information on network_file. 
In model.py, this is how network_file is set up.
```
class VirusOnNetwork(mesa.Model):
    """A virus model with some number of agents"""

    def __init__(self,
                 network_file,
                 initial_outbreak_size=1,
                 virus_spread_chance=0.4,
                 virus_check_frequency=0.4,
                 recovery_chance=0.3,
                 gain_resistance_chance=0.5):
        
        self.G = nx.read_edgelist(network_file, delimiter= " ", nodetype = int)
        self.num_nodes = len(self.G.nodes())
        self.schedule = RandomActivation(self)
        self.grid = mesa.space.NetworkGrid(self.G)
        self.schedule = mesa.time.RandomActivation(self)
etc...
```
As for information about the actual .edge file, it has 333 nodes and 2519 edges. However, when I tried reading in another .edge file with only 3 nodes and 3 edges (i.e. a triangle), it still does not show up as a network correctly. 

I wonder if it's the limitations of the visualization module itself. With the original example model, it can handle up to 400 nodes, which is over the number of nodes I have, but is it possible that the number of edges is the cause of it not rendering?

**Reply by rht** [reply]
That's useful info. Let's simplify the situation to 3 nodes and 3 edges for now.

- What if you set `network_file=""` as a keyword argument?
- You can see the console message if you are on Chromium-based browser, by pressing ctrl-shift-i, or cmd-shift-i (on macOS)

**Reply by jenny61599** [reply]
Thank you for the suggestion, this is what my updated __init__ looks like:
```
class VirusOnNetwork(mesa.Model):
    """A virus model with some number of agents"""

    def __init__(self,
                 network_file = "",
                 initial_outbreak_size=1,
                 virus_spread_chance=0.4,
                 virus_check_frequency=0.4,
                 recovery_chance=0.3,
                 gain_resistance_chance=0.5):
```
The .txt (.edge) file for 3 nodes is formatted as follows:
```
1 2
2 3
3 1
```
As for the console messages, when I run the server using the simplified 3-node file, this is what shows up:

![Screenshot 2023-03-16 065639](https://user-images.githubusercontent.com/81499095/225640514-cc37771d-670e-4778-b5d1-84129882b07e.png)

And the same error shows up for the original .edge file with all 333 nodes too:

![Screenshot 2023-03-16 065741](https://user-images.githubusercontent.com/81499095/225640795-3882b3c4-ca81-4d77-83c7-72a15d38c772.png)

**Reply by rht** [reply]
Can you try to index the node starting from 0 instead of 1? This is a limitation of D3.js.

**Reply by jenny61599** [reply]
I looked at my code in server.py and model.py, but I don't think I indexed the nodes starting from 1. When I print out the portrayal of the 3 node file, it does return the correct `portrayal["nodes"]` and `portrayal["edges"]`.
```
[{'size': 6, 'color': '#FF0000', 'tooltip': 'id: 1<br>state: INFECTED'}, {'size': 6, 'color': '#008000', 'tooltip': 'id: 2<br>state: SUSCEPTIBLE'}, {'size': 6, 'color': '#008000', 'tooltip': 'id: 3<br>state: SUSCEPTIBLE'}]
[{'source': 1, 'target': 2, 'color': '#e8e8e8', 'width': 2}, {'source': 1, 'target': 3, 'color': '#e8e8e8', 'width': 2}, {'source': 2, 'target': 3, 'color': '#e8e8e8', 'width': 2}]
```
Furthermore, when I click on "step" in the browser, it appears that portrayal is updated to reflect the spread of infection. (See how nodes 2 and 3 were susceptible and now they are infected)
```
{"type":"get_step","step":1}
[{'size': 6, 'color': '#FF0000', 'tooltip': 'id: 1<br>state: INFECTED'}, {'size': 6, 'color': '#FF0000', 'tooltip': 'id: 2<br>state: INFECTED'}, {'size': 6, 'color': '#FF0000', 'tooltip': 'id: 3<br>state: INFECTED'}]
[{'source': 1, 'target': 2, 'color': '#e8e8e8', 'width': 2}, {'source': 1, 'target': 3, 'color': '#e8e8e8', 'width': 2}, {'source': 2, 'target': 3, 'color': '#e8e8e8', 'width': 2}]
```

**Reply by rht** [reply]
You have to start the node index from 0, and the edge list needs to reflect that as well. The simulation works. The problem is in the visualization engine (D3.js) that is inflexible.

**Reply by jenny61599** [reply]
Thank you for the clarification. I'm still not very familiar with Python (this is my first project!), so apologies if I am slower at understanding things. Would the node index issue be in the following code from server.py? Also, what should the node and edge list look like? 
```
def get_agents(source, target):
        return G.nodes[source]["agent"][0], G.nodes[target]["agent"][0]

    portrayal = dict()
    portrayal["nodes"] = [
        {
            "size": 6,
            "color": node_color(agents[0]),
            "tooltip": f"id: {agents[0].unique_id}<br>state: {agents[0].state.name}",
        }
        for (_, agents) in G.nodes.data("agent")
    ]

    portrayal["edges"] = [
        {
            "source": source,
            "target": target,
            "color": edge_color(*get_agents(source, target)),
            "width": edge_width(*get_agents(source, target)),
        }
        for (source, target) in G.edges
    ]

    return portrayal
```

**Reply by rht** [reply]
I think the solution with the fewest lines of code would be to decrement all the from-id's and to-id's when you read from the edge list. Everything else doesn't need to be modified.

**Reply by jenny61599** [reply]
Hey, I actually ended up fixing it yesterday with another lab member! I'll share our solution but please let me know if your way is still applicable (for other people just in case they stumble on this thread in the future).

 1) After adding `0 0` to our 3-node file, the network did should up in the visualization with one node that is away from everyone else. However, when we added the same thing to the big file, the graph still did not render. So we investigated the difference between the 3-node file and the original file. 

2) The core issue is that the original file had only 333 nodes but the maximum node number was around 347, which means some nodes were missing. We created 333 agents initially with 333 nodes. What fixed it was adding in those missing nodes to the networkx object first, then creating 347 agents, and placing those agents on the nodes.

See the code below
```
        self.G = nx.read_edgelist(network_file, delimiter= " ", nodetype = int)
        self.num_nodes = max(self.G.nodes) + 1
        self.schedule = RandomActivation(self)

        # add all the possible node values to G before it converts to a mesa network object
        self.G.add_nodes_from(range(self.num_nodes))
        self.grid = mesa.space.NetworkGrid(self.G)
        self.schedule = mesa.time.RandomActivation(self)
```
3) After doing this, the network showed up on the server! 

Thank you so much for your help, but again, please let me know if there was a simpler way of doing it. Most importantly, is it possible to add/delete nodes and edges in the NetworkGrid object? I can imagine a simulation where agent interactions can form or break an edge (maybe also based on spatial distance), but I'm not sure if it's possible with the visual limitations.

**Reply by rht** [reply, marked as answer]
Your solution overcounts the number of nodes by 1 (node 0 that is never used). This may not affect the result of your analysis (if you do some network measures) since the number of nodes is off by 0.2% (but 25% for the 3-node case).

As with the 2nd situation, this means that there are nodes without edges from the edge list.

In both situations, they can be solved if you pre-process the edge list file by decrementing the numbers by 1 before you read it. Then you will need only 333 nodes.
